### ISSUE TITLE
importance_sampling_level=None (the default) gives GRPO an exactly-zero policy gradient — weight-level evidence, and a respectful follow-up to #50 (Bug 1)

### ISSUE THREAD

**Issue by swvgjbb78s-coder** [Goekdeniz-Guelmez/mlx-lm-lora#55]
## Issue title

`importance_sampling_level=None` (the default) gives GRPO an exactly-zero policy gradient — weight-level evidence, and a respectful follow-up to #50 (Bug 1)

### TL;DR, in plain language

If you run `--train-mode grpo` **starting from a base model** and don't explicitly set
`importance_sampling_level`, your model does not learn. Not slowly — not at all. The rewards compute and
print real means and stds, the loss shows plausible values, the adapters save on schedule — but the policy
gradient is exactly zero by construction, and the LoRA B-matrices never leave their zero initialization. The
saved "trained" adapter is arithmetically the base model. One yaml line (`importance_sampling_level: token`)
restores real training. Runs **resumed from an existing adapter** are subtler: their weights do move, but
only from KL-drift back toward the reference — the reward never reaches them either.

### Context — who I am and how I found this

I'm an independent researcher fine-tuning Llama-3.2-3B (4-bit) with GRPO and custom LLM-judge reward
functions, for a narrative-comprehension task — running entirely on an **8 GB MacBook Air**. This package is
what makes that possible at all (to my knowledge nothing else offers GRPO with a custom-reward registry on
Apple Silicon at this footprint), so this report comes with genuine gratitude: I depend on it.

Because my reward is judge-based (an LLM scores answer quality) rather than verifiable (no unit test or exact
answer exists for comprehension), I'd built unusually careful instrumentation around the loop — a validated
evaluator with positive/negative controls, and a habit of checking artifacts rather than dashboards. When a
fully-instrumented 60-iteration run came back statistically identical to the base model, that instrumentation
let me eliminate the reward, the judge, and the eval as suspects — which left only the weights. I loaded the
adapters and looked: `lora_b` was exactly zero. Working backward from there led to the same line of code #50
found in April.

### Relation to #50 — why I'm respectfully re-raising Bug 1

(For readers landing here cold: #49 was an April 2026 pull request proposing a fix; #50 is its companion bug
report; both were closed after discussion.) I've read #50 and #49 in full, including the design rationale
("the sampling level is set to none to replicate the OG paper") and the filer's retraction. I arrived at Bug
1 independently, with a kind of evidence that thread didn't have: **weight forensics and a controlled
one-line A/B experiment.** I believe the retraction on Bug 1 was premature — the filer's other claims may
well have been mistaken, but this one was correct — and I'd like to make the case carefully: mechanism,
evidence, then a fix that *preserves* the paper-replication intent.

### The mechanism (v2.1.0, `mlx_lm_lora/trainer/grpo_trainer.py` line 441)

```python
elif importance_sampling_level is None or importance_sampling_level == "none":
    log_importance_weights = mx.zeros_like(log_ratio)
```

`mx.zeros_like` creates a **new constant array with no connection to the computation graph**. Downstream:

```python
coef_1 = mx.exp(log_importance_weights)             # == 1.0, constant, NO gradient path
unclipped_obj = coef_1 * advantages.reshape(-1, 1)  # constant × constant (advantages carry no grad)
per_token_loss = -mx.minimum(unclipped_obj, clipped_obj)  # constant w.r.t. model parameters
```

So under the default, the entire policy objective contributes **exactly zero gradient** — not small, zero.
The only gradient left is the KL term, whose per-token gradient is proportional to `(ratio − 1)` where
`ratio = π_θ / π_ref`:

- **Fresh-from-base runs:** policy ≡ reference at init → ratio ≡ 1 → the KL gradient is also exactly zero →
  **the total gradient is exactly zero for the entire run.** The model cannot train, whatever the reward.
- **Resumed-from-adapter runs:** policy ≠ reference → the KL gradient is nonzero, but it only pulls the
  policy back **toward the reference**. Weights move — but the movement is reward-free drift. The reward is
  computed, logged, and multiplied by a constant.

### Why this is easy to miss (the symptom fingerprint)

Everything *looks* alive. Rewards print with real means and stds. The progress bar advances. Loss values are
plausible: with equal-length completions the reported loss pins at `0.000` (group-normalized advantages sum
to zero), and with variable-length completions it shows nonzero values (token-weighted advantage means) —
**loss values without gradient**. I ran two full 60-iteration runs against this and diagnosed my results for
a day before checking the weights.

### The decisive evidence (all reproducible)

1. **LoRA B-matrices stay at exactly 0.0.** (LoRA trains small adapter pairs A×B instead of the full model;
   `lora_b` initializes to zero by design — so it moves if and only if training does anything.) After two
   independent 60-iteration fresh-from-base runs (different data, different reward functions, both with
   verified within-group reward spread on most steps — σ up to 0.75 in the logs), `max(abs(lora_b)) == 0.0`
   exactly, across every adapted layer, in both runs.
2. **The controlled flip.** Changing **only** `importance_sampling_level: token` in the config — same model,
   same data, same reward, same seed handling — moves `max(abs(lora_b))` off zero within 3 iterations
   (1.2e-04), and to ~2.7e-03 over a 60-iteration run. One line separates "trains" from "provably cannot."
3. **The resumed-run signature.** A run resumed from an SFT adapter under the default moved all 112 adapter
   tensors by small amounts (~1e-4 scale) — consistent with uniform KL-pull toward the reference and
   inconsistent with reward-driven learning. This is where the default deceives hardest: weights change,
   metrics wiggle, and none of it is the reward.

Falsifiability, stated plainly: if the default is a functioning training mode, a 3-iteration run with any
spread-producing reward should move `lora_b` off exact zero. It does not.

### On the design rationale — respectfully

I understand the intent in #50: `None` replicates the original GRPO formulation (no importance-sampling
correction). I think there's a subtle conflation between two different things:

- **"No importance-sampling correction"** (the paper-faithful reading): the objective becomes plain
  advantage-weighted log-likelihood — REINFORCE-style, `loss = -(advantages * token_log_probs)` — which has a
  real gradient and trains. Notably, the `"token"` mode at ratio = 1 (on-policy) has **exactly this
  gradient** via the standard `exp(logp − stop_gradient(logp_ref))` construction.
- **What the `zeros_like` branch implements:** no gradient through the policy term at all. No published GRPO
  variant — including the original DeepSeekMath formulation — trains with a zero policy gradient. The
  original paper's objective (Eq. 3, arXiv:2402.03300 §4.1.1) is exactly
  `min(π_θ/π_θold · Â, clip(π_θ/π_θold, 1−ε, 1+ε) · Â) − β·D_KL[π_θ‖π_ref]` — the ratio-times-advantage term
  is fundamental to it, and the paper defines no ratio-free variant.

So honoring the design intent actually *requires* a code change: either `None` should implement the
REINFORCE-style objective, or the default should move to `"token"` (gradient-equivalent on-policy).

### Suggested remedies (any subset; happy to PR whichever you prefer)

1. **Change the default** to `"token"` (a default that trains matches user expectation), and/or
2. **Implement `"none"` as true no-IS REINFORCE** (`-advantages * token_log_probs`) — the paper-faithful
   semantics the current branch intends, and/or
3. **Loud guard**: if the level resolves to `None/"none"` under current semantics, emit a prominent warning
   ("policy-gradient term carries no gradient; training proceeds via KL only — fresh-from-base runs will not
   learn"), and/or
4. A **post-run sanity helper**: warn if all `lora_b` remain exactly zero after training. This check is what
   finally caught it for me, and it would catch several failure classes beyond this one.

### Minimal reproduction (~10 minutes, no cloud, no custom data beyond a tiny prompt file)

```python
# rewards_repro.py — deterministic reward with guaranteed within-group spread (no external calls)
from mlx_lm_lora.trainer.grpo_reward_functions import register_reward_function

@register_reward_function("spread_probe")
def spread_probe(prompts, completions, answer=None, types=None):
    return [((len(c or "") % 29) / 29.0) - 0.5 for c in completions]
```

```bash
# Arm A — default (importance_sampling_level unset). Prediction: lora_b stays exactly 0.
python -m mlx_lm_lora.train --model mlx-community/Llama-3.2-3B-Instruct-4bit \
  --train --train-mode grpo --train-type lora --num-layers 8 --batch-size 1 --group-size 6 \
  --temperature 1.0 --iters 3 --learning-rate 1e-5 \
  --reward-functions-file rewards_repro.py --reward-functions spread_probe \
  --data <any small prompt-only dataset> --adapter-path ./adapter_none --max-completion-length 32

# Arm B — identical + a yaml containing:  importance_sampling_level: token
# Prediction: lora_b moves off zero within 3 iterations.
```

```python
# check_b.py — the assertion
import mlx.core as mx
w = mx.load("adapter_none/adapters.safetensors")
print(max(float(mx.abs(v).max()) for k, v in w.items() if "lora_b" in k.lower()))
# 0.0 under arm A; > 0 under arm B
```

Environment: macOS (Apple Silicon, 8 GB), Python 3.12, `mlx` 0.31.2, `mlx-lm` 0.31.3, `mlx-lm-lora` 2.1.0.

Again — thank you for building and maintaining this. The package made my whole research direction possible on
consumer hardware; this report exists because I depend on it. Happy to provide full logs, the two
60-iteration adapters, or a PR for any remedy above.

**Comment by Goekdeniz-Guelmez** [comment]
Hey, thanks for flagging this, fix coming shortly

**Comment by Goekdeniz-Guelmez** [comment]
@swvgjbb78s-coder fix is merged, the new version will be released later this week!
